api/src/main/java/org/openmrs/module/pihcore/task/PihCloseStaleVisitsTask.java

```diff
@@ -54,13 +55,65 @@ public void execute() {
 
         List<Visit> openVisits = visitService.getVisits(null, null, locations, null, null, null, null, null, null, false, false);
         for (Visit visit : openVisits) {
-            if (!changedOrUpdatedRecently(visit, REGULAR_VISIT_EXPIRE_TIME_IN_HOURS) &&
-                (
-                    isOldEDVisit(adtService.wrap(visit), ED_VISIT_EXPIRE_VERY_OLD_TIME_IN_HOURS)
-                    || (adtService.shouldBeClosed(visit) && !isActiveEDVisit(adtService.wrap(visit), ED_VISIT_EXPIRE_TIME_IN_HOURS))
-                    || wasDischargedAndNotAdmitted(adtService.wrap(visit), REGULAR_VISIT_EXPIRE_TIME_IN_HOURS)
-                )
-            ) {
+
+            VisitDomainWrapper wrappedVisit = adtService.wrap(visit);
+            Boolean changedOrUpdatedRecently = changedOrUpdatedRecently(visit, REGULAR_VISIT_EXPIRE_TIME_IN_HOURS);
+
+            Disposition mostRecentDisposition = wrappedVisit.getMostRecentDisposition();
+            Long hoursSinceLastEncounter = wrappedVisit.getMostRecentEncounter() != null ?
+                    new Duration(new DateTime(wrappedVisit.getMostRecentEncounter().getEncounterDatetime()), new DateTime()).getStandardHours()
+                    : null;
+
+            // **this logic is intentionally verbose, since it can be difficult to follow**
+            Boolean closeVisit = false;
```

Comment:
Nice way to approach this. I can't fit the whole thing in my brain but I trust each line, more or less. If you're uneasy at all about the mutable state variable `closeVisit` you could get rid of it by factoring this whole thing out into a function and making each of those `closeVisit = x` lines into `return x`.